x-pack/plugins/security/server/authentication/authentication_service.ts

```diff
@@ -187,55 +187,71 @@ export class AuthenticationService {
     });
 
     http.registerOnPreResponse(async (request, preResponse, toolkit) => {
-      if (preResponse.statusCode !== 401 || !canRedirectRequest(request)) {
-        return toolkit.next();
-      }
-
       if (!this.authenticator) {
         // Core doesn't allow returning error here.
         this.logger.error('Authentication sub-system is not fully initialized yet.');
         return toolkit.next();
       }
-
       // If users can eventually re-login we want to redirect them directly to the page they tried
       // to access initially, but we only want to do that for routes that aren't part of the various
       // authentication flows that wouldn't make any sense after successful authentication.
-      const originalURL = !request.route.options.tags.includes(ROUTE_TAG_AUTH_FLOW)
-        ? this.authenticator.getRequestOriginalURL(request)
-        : `${http.basePath.get(request)}/`;
-      if (!isLoginPageAvailable) {
-        const customBrandingValue = await customBranding.getBrandingFor(request, {
-          unauthenticated: true,
-        });
+
+      const originalURL = request.route.options.tags.includes(ROUTE_TAG_AUTH_FLOW)
+        ? `${http.basePath.get(request)}/`
+        : this.authenticator.getRequestOriginalURL(request);
+
+      const redirectLoginLogout = async () => {
+        const needsToLogout = (await this.session?.getSID(request)) !== undefined;
+        if (needsToLogout) {
+          this.logger.warn(
+            'Could not authenticate user with the existing session. Forcing logout.'
+          );
+        }
+
         return toolkit.render({
-          body: renderUnauthenticatedPage({
-            basePath: http.basePath,
-            staticAssets: http.staticAssets,
-            originalURL,
-            customBranding: customBrandingValue,
-          }),
-          headers: { 'Content-Security-Policy': http.csp.header },
+          body: '<div/>',
+          headers: {
+            'Content-Security-Policy': http.csp.header,
+            Refresh: `0;url=${http.basePath.prepend(
+              `${
+                needsToLogout ? '/logout' : '/login'
+              }?msg=UNAUTHENTICATED&${NEXT_URL_QUERY_STRING_PARAMETER}=${encodeURIComponent(
+                originalURL
+              )}`
+            )}`,
+          },
         });
+      };
+
+      if (
```

Comment:
__note:__ I think this complex condition warrants a code comment explaining what we're actually trying to achieve here.

**But** I'm wondering if it'd be much easier to understand this function if we structure this function like this instead (maybe adding tests before changes would be safer):
```typescript
// AT THE VERY TOP

// Comment about why we don't care about <400 and non-redirectable requests.
if (preResponse.statusCode < 400 || !canRedirectRequest(request)) {
  return toolkit.next();
}

// Comment about why we don't care about non-401 errors for non-auth flows.
if (preResponse.statusCode !== 401 && !request.route.options.tags.includes(ROUTE_TAG_AUTH_FLOW)) {
  return toolkit.next();
}

// At this point we're only dealing with either errors in auth flow
// or 401 errors in non auth flows - both require some type of UI
// with error - either login page/selector or custom error UI
// Then, we might not need dedicated `redirectLoginLogout` function

const originalURL = ...;
if (!isLoginPageAvailable) {
  ...
  return toolkit.render({ body: renderUnauthenticatedPage({...}) });
}

const needsToLogout = (await this.session?.getSID(request)) !== undefined;
....
return toolkit.render({
  body: '<div/>',
  headers: {
    Refresh: `0;url=${http.basePath.prepend(
      `${
        needsToLogout ? '/logout' : '/login'
      }?msg=UNAUTHENTICATED&${NEXT_URL_QUERY_STRING_PARAMETER}=${encodeURIComponent(
        originalURL
      )}`
    )}`,
  },
});
```

It'd also be easier to write unit tests for this function, just going through and covering all conditions from top to bottom.

What do you think? Did I miss something?